_posts/2021-08-20-A query, or There and Back Again.md

```diff
@@ -0,0 +1,67 @@
+
+### Query Phase
+In this phase, the query provided to OpenSearch is broadcasted to a copy of every shard across the entire index. Once received, the query is executed locally. The result is a priority queue of matching, sorted documents for each shard. This priority queue is simply a sorted list of the *top n* matching documents with *top* being determined by relevance and *n* being determined by pagination parameters set by the user (or the default if not set by the user). Relevance in this case is a score of how well each document it matches the query. The individual shards are responsible for the actual matching process as well as the scoring. So for example, if you have a three node cluster and wanted to search for "Hamster", you could write a query like this;
+```
+curl -X GET "localhost:9200/_search?q=Hamster&pretty"
+```
+Once OpenSearch has your query, OpenSearch takes the following steps;
+1. The API or client sends your search query to Node 1. 
+2. Node 1 sends the search request to a primary or replica shard for each shard in an index. 
+3. Each shard executes the search locally and creates a locally sorted queue.
+4. Each shard returns doc ID and sort value of the all the documents in its local queue.
+5. Node 1 as the coordinator node, merges these values into a globally sorted list.
+
+*Figure 1: Query Phase*
+![Figure 1: Query Phase Diagram](https://user-images.githubusercontent.com/9532728/130289482-4ef3379a-1dbf-49b7-af36-ba9339864005.png)
+
+
+*Note on coordinating nodes*: Once OpenSearch learns about the query, the query can be sent to any available data node and that node can become the coordinating node for that query. These nodes delegate client requests to the shards on the data nodes, collects and aggregates the results into one final result, and sends this result back to the client. Often larger clusters will have dedicated coordinating nodes that manage search volume. 
+
+
+### Fetch Phase
+Now that the query phase has identified the documents that satisfy the request, OpenSearch needs to actually retrieve the documents. For the fetch phase, the coordinating node used the globally sorted priority list generated in the query phase to build the GET requests needed for the query. Using the same three node cluster as the earlier example, OpenSearch needs to collect all the results for your query (i.e. "Hamster") and return them to you. 
+
+1. The coordinating node uses the global list to identify which documents are needed.
+2. The coordinating node issues multiple GET requests to the relevant shards.
+3. Each individual shard loads the request document and returns them to the coordinating node.
+4. Once all the documents are returned, the results are returned to the client.
+
+*Figure 2: Fetch Phase*
+![FetchPhase](https://user-images.githubusercontent.com/9532728/130289580-89b6d8e1-2943-40ae-b547-1858de9a55f2.png)
```

Comment:
See above comment on github hosting and img-fluid tagging.

Reply:
Updated to;

![Figure 2: Fetch Phase Diagram]({{ site.baseurl }}/assets/media/blog-images/2021-08-23-a-query-or-there-and-back-again/FetchPhase.png){: .img-fluid }

Working to validate on local.